### PR TITLE
#11985 Fixes to MDLog (and MDS) shutdown

### PR DESCRIPTION
**Pull request by jcsp**

### CODE REVIEW

**Comment by jcsp** [comment]
NB has a couple of commits from wip-damaged-fixes because both touch journaler and I want to avoid merge pain

**Comment by jcsp** [comment]
http://pulpito.front.sepia.ceph.com/john-2015-06-18_11:24:49-fs-wip-11985-testing-basic-multi/

File: src/osdc/Journaler.cc [hunk 1]
```diff
@@ -1367,3 +1407,30 @@ C_OnFinisher *Journaler::wrap_finisher(Context *c)
...
+
+  // Kick out anyone reading from journal
+  error = -EAGAIN;
+  if (on_readable) {
+    C_OnFinisher *f = on_readable;
+    on_readable = 0;
+    f->complete(-EAGAIN);
+  }
+
+  finish_contexts(cct, waitfor_recover, 0);
```
Comment (gregsfortytwo) [inline src/osdc/Journaler.cc]: Shouldn't this be EAGAIN instead of 0?

**Comment by gregsfortytwo** [comment]
Please review this, @ukernel.

**Comment by ukernel** [comment]
reviewed-by

**Comment by ukernel** [comment]
@jcsp please rebase

**Comment by jcsp** [comment]
@ukernel the test run that I linked earlier had failures in it!  Will fix in another PR, but would have been better not to merge this already

**Pull request merged**